packages/nimble-components/src/rich-text-viewer/index.ts

```diff
@@ -11,7 +17,89 @@ declare global {
 /**
  * A nimble styled rich text viewer
  */
-export class RichTextViewer extends FoundationElement {}
+export class RichTextViewer extends FoundationElement {
+    /**
+     * @public
+     * @remarks
+     * Accessor: uses the parsing logic to convert the input markdown string to render in the component.
+     */
+    public set markdown(value: string) {
+        this._markdown = value;
+        this.serializedContent = this.parseMarkdownToDOM(value);
+        this.updateViewerNodeWithSerializedContent();
+    }
+
+    /**
+     * @public
+     * @remarks
+     * Accessor: gets the raw markdown string.
+     */
+    public get markdown(): string {
+        return this._markdown;
+    }
+
+    private _markdown = '';
+    private serializedContent?: HTMLElement | DocumentFragment;
+    private readonly markdownParser: MarkdownParser;
+    private readonly domSerializer: DOMSerializer;
+
+    public constructor() {
+        super();
+        this.domSerializer = DOMSerializer.fromSchema(schema);
+        this.markdownParser = this.initializeMarkdownParser();
+    }
+
+    /**
+     * @internal
+     */
+    public override connectedCallback(): void {
+        super.connectedCallback();
+        this.updateViewerNodeWithSerializedContent();
+    }
+
+    private initializeMarkdownParser(): MarkdownParser {
+        /**
+         * It configures the tokenizer of the default Markdown parser with the 'zero' preset.
+         * The 'zero' preset is a configuration with no rules enabled by default to selectively enable specific rules.
+         * https://github.com/markdown-it/markdown-it/blob/2b6cac25823af011ff3bc7628bc9b06e483c5a08/lib/presets/zero.js#L1
+         *
+         */
+        const zeroTokenizerConfiguration = defaultMarkdownParser.tokenizer.configure('zero');
+
+        // The detailed information of the supported rules were provided in the below CommonMark spec document.
+        // https://spec.commonmark.org/0.30/
+        const supportedTokenizerRules = zeroTokenizerConfiguration.enable([
+            'emphasis',
+            'list',
+            'autolink'
+        ]);
+
+        return new MarkdownParser(
+            schema,
+            supportedTokenizerRules,
+            defaultMarkdownParser.tokens
+        );
+    }
+
+    /**
+     *
+     * This function takes a markdown string, parses it using the ProseMirror MarkdownParser, serializes the parsed content into a
+     * DOM structure using a DOMSerializer, and returns the serialized result.
+     *
+     */
+    private parseMarkdownToDOM(value: string): HTMLElement | DocumentFragment {
+        return this.domSerializer.serializeFragment(
+            this.markdownParser.parse(value)!.content
+        );
+    }
+
+    private updateViewerNodeWithSerializedContent(): void {
+        const viewer = this.shadowRoot?.querySelector('#viewer');
+        if (viewer && this.serializedContent) {
+            viewer.replaceChildren(this.serializedContent.cloneNode(true));
```

Comment:
Interested to hear others thoughts on this, but would we prefer an implementation that didn't mess with the DOM imperatively, and instead used `XMLSerializer` to get us the string representation of the DOM, and then just bind the template to an observable using that serialized string? It would look like:

```ts
    private updateViewerNodeWithSerializedContent(): void {
        if (this.serializedContent) {
            this.viewerContents = html`${this.xmlSerializer.serializeToString(this.serializedContent)}`;
        }
    }
```

template code:
```html
    <template>
        <div class="container" id="viewer">
            ${x => x.viewerContents}
        </div>
    </template>
```

Reply:
@atmgrifter00 nice suggestion. I personally like the direction of moving more DOM structure to the template. I agree with your pro and also like that it delegates management of that DOM to FAST (I don't know the details of what FAST does but I'd rather trust them to do the right validation, caching, etc than think about that ourselves). To be safe we should probably re-validate performance with that approach, but I doubt it'll be a problem since the markdown doesn't change frequently in the viewer component.

If it doesn't work out then a different suggestion for this function is not to look up the parent element every call using `const viewer = this.shadowRoot?.querySelector('.viewer');`. Instead you can use FAST's `ref` directive [to get an HTML element reference](https://www.fast.design/docs/fast-element/using-directives#referential-directives) to the element in this class.

Reply:
Agree with the premise that we shouldn't minimize how much we imperatively mess with the DOM. The idea of reserializing the node tree back to a string and re-parsing it seems expensive. I think we can move a couple things around to make this more idiomatic FAST.

```ts
export class RichTextViewer extends FoundationElement {

@observable
public markdown?: string;
public viewer: HTMLDivElement;

connectedCallback() {
    super.connectedCallback();
    this.updateViewer();
}

markdownChanged() {
  // Avoid the extra work until the DOM is rendered
   if (this.$fastController.isConnected) {
       this.updateViewer();
   }
}

updateViewer () {
     const serializedContent = // parse this.markdown
     this.viewer.replaceChildren(serializedContent); // Can we avoid the cloneNode? Not sure why we need a copy of the nodes
}
```

Template:

// Note in the template you can leave off the surrounding `<template>` if you don't need to bind to the host (i.e. modify the `nimble-rich-text-viewer` attributes itself)
```ts
html<RichTextViewer >`<div ${ref('viewer')} class="viewer"></div>`;
```

Takeaways:
- We can use observables and `*Changed` to avoid the accessor in this case to have a property based API since it's a dumb property assignment and read
- Can avoid doing the work until we need to, i.e. only when the DOM is actually constructed after connect.

Reply:
I have updated `markdown` to have a property based API  instead of an accessor and rendered the component using FAST's `ref` directive as suggested.